### ISSUE TITLE
Show when upgrade is in_progress (like units)

### ISSUE THREAD

**Issue by MetriC-DT** [ZephyrBlu/zephyrus-sc2-parser#21]

**Comment by ZephyrBlu** [comment]
At the moment upgrade information is stored in `Player` objects (`Player.upgrades`) and in the timeline (`timeline[<player_id>]['upgrade']`).

I can easily add the start time to the information in `Player` objects, but because the upgrade information in the timeline is a list, adding progress information would be a breaking change.

A crude solution would be to add a `upgrade_in_progress` key to the timeline containing a list of upgrades in progress like so: (`timeline[<player_id>]['upgrade_in_progress']`).

Is there something specific you want to use this data for? I could add more information like when the upgrade was started if that's helpful.

**Comment by MetriC-DT** [comment]
Yeah, something like when the upgrade is started is helpful. It's for the fact that build orders times are generally displayed when the unit, building, or upgrade is first started, so having this information would be nice.

**Comment by ZephyrBlu** [comment]
Ok, I'll create a POC today and link the PR here.

For buildings the time they're started at is easy, it's `GameObj.init_time`.

For units things can be a bit trickier. Presumably you want the time the unit was queued rather than the time it started training, but that's not the easiest information to get.

One way to kind of get around this would be to group unit train times by the first one.

Ex: 5x Stalkers @5min would mean the first Stalker was started at the 5min mark and the others were continuously trained after that.

Let me know about the unit stuff and I might be able to help out, since I don't think that info (Train times) is currently accessible for Terran and Zerg units :).

**Comment by MetriC-DT** [comment]
Sorry if I'm misunderstanding something, but isn't that information already exposed with the units' "in_progress" key? If we add up "in_progress" + "alive" + "dead" and check the difference between the current and previous timeline index, can't we determine the time a unit is started? (this is the way I'm checking for unit queued in my program)

**Comment by ZephyrBlu** [comment]
Ah, I see how you would think that haha. Thinking about it now, the `in_progress` key in the timeline is quite misleading.

This is the only place I currently modify object status to `IN_PROGRESS` is here:

https://github.com/ZephyrBlu/zephyrus-sc2-parser/blob/fdcc298ced9f2f8194d4c864ef2943382214c0bd/zephyrus_sc2_parser/events/object_event.py#L127-L136

For context, this event type (`SUnitInitEvent`) only occurs for buildings and warp-ins so non-warped in units will never have the `IN_PROGRESS` status.

Object statuses are serialized to the timeline here:

https://github.com/ZephyrBlu/zephyrus-sc2-parser/blob/fdcc298ced9f2f8194d4c864ef2943382214c0bd/zephyrus_sc2_parser/game/player_state.py#L126-L127

So non-warped in units will never be in progress on the timeline. If I could add that information to the timeline I would, but the information isn't directly in the replay events.

---

I can kind of reconstruct the information by backtracking from the birth time of units though. You can find the reconstructed information in `Player.queues` which contains units that are currently being trained.

I don't think I've updated the README with building queues yet, but it's pretty simple. `Player.queues` is a list of queue states for that player. Each queue state looks like this:

```
{
    'gameloop': <gameloop>,
    'supply_blocked': <boolean>,
    'queues': {<building>: <dequeue list>},
    'downtime': {'current': <current downtime>, 'total': <total downtime>},
}
```

Queue states are added when there's a change in either player's queues, so sometimes there will be duplicate states (With different gameloops) in each player's list of queue states. 

The `queues` key is where the relevant information is. I didn't realize until I'd already built it out, but there's only ever one unit in each building's queue (Unless Reactors make something weird happen).

You should be able to aggregate all building queues to get all the units currently being created. There might be a few edge cases that I haven't ironed out, but if you run into one I'm happy to fix it.

---

Btw, if you want exact timings for buildings I suggest filtering through `Player.objects.values()` instead of the timeline like so:

```
for player in replay.players.values():
    for obj in player.objects.values():
        if 'BUILDING' in obj.type:
            # do something with the building
            # obj.init_time is the time the building was started
```

Thinking about this, it probably makes more sense to just add the start time to this upgrade class:

https://github.com/ZephyrBlu/zephyrus-sc2-parser/blob/fdcc298ced9f2f8194d4c864ef2943382214c0bd/zephyrus_sc2_parser/dataclasses.py#L24-L28

That way you can access the exact time all upgrades were started through `Player.upgrades`.

**Comment by ZephyrBlu** [comment]
Super simple starting point for the PR: https://github.com/ZephyrBlu/zephyrus-sc2-parser/pull/22/files

I still need to add the research times for this code to work though, and this implementation won't deal with edge cases like Chronoboost (That's probably doable but more complicated.
